src/liballoc/boxed.rs

```diff
@@ -414,3 +416,69 @@ impl<'a,A,R> FnOnce<A> for Box<FnBox<A,Output=R>+Send+'a> {
 }
 
 impl<T: ?Sized+Unsize<U>, U: ?Sized> CoerceUnsized<Box<U>> for Box<T> {}
+
+impl<T: Clone> Clone for Box<[T]> {
+    fn clone(&self) -> Self {
+        let mut alloc = unsafe {
+            heap::allocate(mem::size_of::<T>() * self.len(),
+                           mem::align_of::<T>()) as *mut T
+        };
+
+        if alloc == ptr::null_mut() {
+             ::oom()
+        }
+
+        let mut new = BoxBuilder {
+            data: alloc,
+            len: 0,
+            capacity: self.len(),
+            phantom: PhantomData
+        };
+
+        for item in self.iter() {
+            unsafe {
+                ptr::write(alloc, item.clone());
+                alloc = alloc.offset(1);
+            };
+
+            new.len += 1;
+        }
+
+        let data = mem::replace(&mut new.data, ptr::null_mut());
+
+        unsafe { mem::transmute(Slice {
+             data: data,
+             len: self.len()
+        }) }
+    }
+}
+
+struct BoxBuilder<T> {
+    data: *mut T,
+    len: usize,
+    capacity: usize,
+    phantom: PhantomData<T>
+}
+
+impl<T> Drop for BoxBuilder<T> {
+    fn drop(&mut self) {
+        if self.data != ptr::null_mut() {
```

Comment:
I don't think the null check is necessary in the first place: it's impossible to construct a BoxBuilder with a null data pointer.